### PR TITLE
Make ghc-source-gen compatible with GHC 9.10

### PR DESCRIPTION
**Pull request by wavewave**
Adjusted with all the new API changes.

### CODE REVIEW

**Comment by blackgnezdo** [comment]
@wavewave thanks for the change!

Do you mind updating the CI system? It's been a while since anybody contributes non-trivial changes, so our CI bitrotted. If you don'h a favorite one, feel free to steal from e.g. google/proto-lens

**Comment by blackgnezdo** [comment]
@jinwoo do you happen to be capable of approving this to get the workflow going?

**Comment by jinwoo** [comment]
@blackgnezdo Looks like I can. But some tests are failing.

File: stack-9.10.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2019 Google LLC
+#
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file or at
+# https://developers.google.com/open-source/licenses/bsd
+
+resolver: nightly-2024-04-06
```
Comment (blackgnezdo) [inline stack-9.10.yaml:7]:
```suggestion
resolver: nightly-2024-10-28
```

Maybe use the newest since we are carving something custom anyway?

**Comment by jinwoo** [comment]
All tests seem to have passed now. @blackgnezdo Do you think it's okay to merge this?

File: src/GHC/SourceGen/Overloaded.hs [hunk 2]
```diff
@@ -185,14 +227,20 @@ instance HasTuple HsExpr' where
         explicitTuple = withEpAnnNotUsed ExplicitTuple
 #else
         explicitTuple = noExt ExplicitTuple . map builtLoc
+#endif
 #endif
     unit = noExt HsVar unitDataConName
 
+
```
Comment (blackgnezdo) [inline src/GHC/SourceGen/Overloaded.hs:234]:
```suggestion
```
Accidental newline?

File: src/GHC/SourceGen/Syntax/Internal.hs [hunk 3]
```diff
@@ -128,7 +131,10 @@ noExtOrPlaceHolder :: (GHC.PlaceHolder -> a) -> a
 noExtOrPlaceHolder = withPlaceHolder
 #endif
 
-#if MIN_VERSION_ghc(9,2,0)
+#if MIN_VERSION_ghc(9,10,0)
+withEpAnnNotUsed :: a -> a --(EpAnn ann -> a) -> a
+withEpAnnNotUsed = id -- ($ noAnn)
```
Comment (blackgnezdo) [inline src/GHC/SourceGen/Syntax/Internal.hs:136]:
```suggestion
withEpAnnNotUsed :: a -> a
withEpAnnNotUsed = id
```
Might as well not replicate the code below in the comments.

File: src/GHC/SourceGen/Decl.hs [hunk 4]
```diff
@@ -281,44 +291,79 @@ instance HasTyFamInst RawInstDecl where
...
+    tyFamInstDecl = TyFamInstDecl []
+    eqn_bndrs = noExt HsOuterImplicit
+#elif MIN_VERSION_ghc(9,2,0)
 tyFamInst name params ty = tyFamInstD
         $ tyFamInstDecl
         $ famEqn
             (typeRdrName name)
             eqn_bndrs
-            (map mkLocated params)
+            params -- (map mkLocated params)
```
[the remark this answers is not in the document]
Comment (wavewave) [reply src/GHC/SourceGen/Decl.hs]: I made this change more minimal. Somehow, I've changed `famEqn` to adjust multiple GHC versions, and then didn't put it back as the final version came out. Thanks!

File: src/GHC/SourceGen/Decl.hs [hunk 5]
```diff
@@ -328,41 +373,84 @@ newOrDataType
...
+                (case newOrData of
+                    NewType -> case conDecls of
+                        [decl] -> NewTypeCon $ mkLocated decl
+                        _ -> error "NewTypeCon with more than one decl"
+                    DataType -> DataTypeCons False (map mkLocated conDecls)
+                )
+                (toHsDeriving $ map mkLocated derivs)
+  where
+    cxt = Nothing
+    toHsDeriving = id
```
[the remark this answers is not in the document]
Comment (wavewave) [reply src/GHC/SourceGen/Decl.hs]: Yup! removed :-) 👍

File: src/GHC/SourceGen/Decl.hs [hunk 6]
```diff
@@ -328,41 +373,84 @@ newOrDataType
...
+            $ noExt HsDataDefn
+                cxt
+                Nothing
+                Nothing
+                (case newOrData of
+                    NewType -> case conDecls of
+                        [decl] -> NewTypeCon $ mkLocated decl
+                        _ -> error "NewTypeCon with more than one decl"
+                    DataType -> DataTypeCons False (map mkLocated conDecls)
+                )
```
[the remark this answers is not in the document]
Comment (wavewave) [reply src/GHC/SourceGen/Decl.hs]: now refactored out the repeating part. Thanks!

**Review by blackgnezdo — APPROVED** [review]
@jinwoo this looks pretty good to me now, WDYT?

**Comment by jinwoo** [comment]
Some tests seem to fail. I'll merge once they are fixed.

**Comment by jinwoo** [comment]
All tests have passed. I'm merging now. Thanks, @wavewave and @blackgnezdo !

**Review by jinwoo — APPROVED** [review]

**Pull request merged**